shell/browser/ui/file_dialog_gtk.cc

```diff
@@ -272,8 +339,25 @@ bool ShowOpenDialogSync(const DialogSettings& settings,
   FileChooserDialog open_dialog(action, settings);
   open_dialog.SetupProperties(settings.properties);
 
-  gtk_widget_show_all(open_dialog.dialog());
-  int response = gtk_dialog_run(GTK_DIALOG(open_dialog.dialog()));
+  if (GTK_IS_WIDGET(open_dialog.dialog())) {
+    gtk_widget_show_all(GTK_WIDGET(open_dialog.dialog()));
+  } else {
+    void (*dl_gtk_native_dialog_show)(void*) = nullptr;
+    g_module_symbol(open_dialog.gtk_module(), "gtk_native_dialog_show",
+                    reinterpret_cast<void**>(&dl_gtk_native_dialog_show));
+    dl_gtk_native_dialog_show(static_cast<void*>(open_dialog.dialog()));
+  }
+
+  int response = 0;
+  if (GTK_IS_WIDGET(open_dialog.dialog())) {
+    response = gtk_dialog_run(GTK_DIALOG(open_dialog.dialog()));
+  } else {
+    int (*dl_gtk_native_dialog_run)(void*) = nullptr;
+    g_module_symbol(open_dialog.gtk_module(), "gtk_native_dialog_run",
+                    reinterpret_cast<void**>(&dl_gtk_native_dialog_run));
+    response =
+        dl_gtk_native_dialog_run(static_cast<void*>(open_dialog.dialog()));
+  }
```

Comment:
```c++
if (GTK_IS_NATIVE_DIALOG(dialog_)) {
   ...
} else {
   ...
}
```

Reply:
Also, this code block exists in two places. We should extract-method and create a helper `int RunFileDialog(void*)`

Reply:
void* over a templated function?

Reply:
Adjusted to take a reference to a FileChooserDialog